### PR TITLE
Add the Pyright type checker

### LINKED ISSUE #5727: Enable static type checking
Motivation
----------
Building on #4455, now that Python 2.7-support is over, enabling a static type checker is easier. This additional tooling will help catch new programmer errors and probably reveal pre-existing bugs. With a proper configuration, the type checker will hopefully not be too noisy, but still provide valuable feedback. Integrating the tooling into the CI pipeline in a similar manner to the `flake8-annotations` check could help catch new errors without complaining too much about old errors.

Modification
------------
The first step is to enable a static type checker with a suitable configuration. There will probably be a lot of typing errors and missing annotations, which can be gradually reviewed and resolved through follow-up PRs. 

The current `flake8-annotations` CI check will still be left in place, since ensuring proper coverage of annotations is vital to also get useful results from the static type checker.

Some pointers for converting an existing codebase with MyPy are provided [here](https://mypy.readthedocs.io/en/stable/existing_code.html).

### PR DESCRIPTION
**Pull request by akselaase**
This PR introduces tooling for the Pyright Python type checker, usable both as a local tool during development, and also integrated into the CI pipeline.

The CI check works in a similar way to the current `flake8-annotations` check, in that it compares the number of typing errors before and after applying an incoming pull request. However, the existing check only counts the number of missing annotations, which does not help in ensuring correctness of the annotations. For instance, [this parameter](https://github.com/rucio/rucio/blob/58efd21b5e21182df80bef3dbe8befa636e440b8/lib/rucio/core/transfer.py#L581) should probably not be a `List[str]`, but a `str`, which Pyright rightly notices in the function body.

The CI check prevents new PRs from increasing the error count. The check also attempts to intelligently recognize where the new errors were introduced, to help the developer find their mistakes more easily. However, this could cause a disruption in other developers' workflow, so maybe some guidelines for how strictly we adhere to the CI result should be developed. Pyright's config can also be modified to be more relaxed if it proves to be too noisy. 

Pyright was chosen due to the relative ease of configuring it and processing its output compared to MyPy. Other issues like MyPy limiting the amount of output in case of too many errors also led to the PR focusing on Pyright (though this issue could maybe be resolved by passing the `--soft-error-limit N` flag to MyPy).

Another benefit of Pyright (that could just as well be present in other type checkers - I didn't look) is the ability to integrate it into your IDE of choice through the Language Server Protocol. I know at least VSCode has seamless integration when using the [Pylance](https://marketplace.visualstudio.com/items?itemName=ms-python.vscode-pylance) extension, which greatly improves the developer experience by highlighting errors immediately, without having to wait for the CI check to run.

As a follow-up to this PR, I want to gradually improve type annotations and correctness throughout the codebase. Pyright reports numerous issues, some of which could be bugs that need fixing.

Closes #5727

### CODE REVIEW

File: tools/run_pyright/__main__.py
Comment (joeldierkes) [inline tools/run_pyright/__main__.py]: I don't see the value of this line, `func` is always initialized.

File: tools/run_pyright/compare.py
Comment (joeldierkes) [inline tools/run_pyright/compare.py]: I only know this argument from gcc, there it is `--Werror`...

File: tools/run_pyright/compare.py
Comment (joeldierkes) [inline tools/run_pyright/compare.py]: `serverity` could be an enum. Thoughts?

File: .github/workflows/autotest.yml
Comment (joeldierkes) [inline .github/workflows/autotest.yml]: I want to see the output of the GitHub Actions job. Should we treat warnings here as errors? (`--werror`)

File: tools/run_pyright/generate.py
Comment (joeldierkes) [inline tools/run_pyright/generate.py]: While I like return codes, I am not completely sure if this one is right. The doc comment says `Generate a pyright-report and save it at the specified path.`, which implies for me that a non-zero return code results in a not created pyright-report, not that pyright found some errors. Thoughts?

File: tools/run_pyright/generate.py
Comment (joeldierkes) [inline tools/run_pyright/generate.py]: An explanation of why we need that would be nice, not necessarily in the git history in that case ;-)

File: tools/run_pyright/models.py
Comment (joeldierkes) [inline tools/run_pyright/models.py]: Number of spaces before comment

File: tools/run_pyright/utils.py
Comment (joeldierkes) [inline tools/run_pyright/utils.py]: `save_json` is called once with a `Path`. Is the recursive call and the split of the parameter type better compared to:
```python
def save_json(path: Path, data: Dict[str, Any]) -> None:
    with open(path, 'w') as f:
        json.dump(data, f, indent=4)
        f.write('\n')
```

File: .github/workflows/autotest.yml
Comment (joeldierkes) [inline .github/workflows/autotest.yml]: Maybe a more descriptive name than the package, like `Check Types`/`Check Static Typing`?

File: tools/run_pyright/generate.py
Comment (joeldierkes) [inline tools/run_pyright/generate.py]: The error message in case `pyright` is not installed is not descriptive:
```
...
  File "/usr/lib/python3.10/subprocess.py", line 1845, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'pyright'
```
I'd suggest to either give more information on how to resolve that, or maybe use `npx`?

**Review by joeldierkes — CHANGES_REQUESTED** [review]
While I like `mypy`, I don't mind using some other tool. In this case some more information would be nice:
- Why don't we use mypy?
- Why pyright? There are other tools like facebooks [`pyre-ceck`](https://github.com/facebook/pyre-check) and googles [`pytype`](https://github.com/google/pytype).

Very good PR, this helps a lot! Small nits/Questions.

File: tools/run_pyright/__main__.py
Comment (akselaase) [reply tools/run_pyright/__main__.py]: It's to type it properly. Ensures that errors show up if any of the `{generate,compare,...}.setup_parser()` calls doesn't return the correct type.

File: .github/workflows/autotest.yml
Comment (akselaase) [reply .github/workflows/autotest.yml]: How about "Python type check (Pyright)"?

File: tools/run_pyright/generate.py
Comment (akselaase) [reply tools/run_pyright/generate.py]: At the moment unused, but I initially used it to produce reports that could be diffed without getting too much noise from changing line numbers. Unsure whether to leave it there or not, thoughts? If so, will explain it better.

File: .github/workflows/autotest.yml
Comment (akselaase) [reply .github/workflows/autotest.yml]: Both reports are now shown in a separate step in the workflow.

File: .github/workflows/autotest.yml
Comment (akselaase) [inline .github/workflows/autotest.yml]: To decide: With or without `--Werror`?

File: .github/actions/checkout_ancestor_commit/action.yaml
Comment (akselaase) [inline .github/actions/checkout_ancestor_commit/action.yaml]: Could also just rename `files_to_copy` to `paths_to_copy` and have it support both files and directories by just adding `-r` to the `cp` command, instead of having separate loops for files and directories.

File: tools/run_pyright/__main__.py
Comment (akselaase) [inline tools/run_pyright/__main__.py]: Not needed when were setting `PYTHONIOENCODING=utf8` I believe, todo: test.

File: tools/run_pyright/generate.py
Comment (akselaase) [inline tools/run_pyright/generate.py]: This doesn't really provide anything useful.

File: .github/actions/checkout_ancestor_commit/action.yaml
Comment (joeldierkes) [reply .github/actions/checkout_ancestor_commit/action.yaml]: That's the cleaner alternative, go for it.
Please create a proper commit with a proper commit message for that (including the intention).

You can force push the branch and amend/edit your changes. The comments on GitHub are still present in the view (if the line is not deleted) and you have revisions for each review.

File: tools/run_pyright/generate.py
Comment (akselaase) [reply tools/run_pyright/generate.py]: Agreed to remove since it is unused.

**Comment by akselaase** [comment]
Demo of output can be seen in the "Compare reports" step here: https://github.com/rucio/rucio/runs/7469292269?check_suite_focus=true

File: tools/run_pyright/compare.py
Comment (joeldierkes) [inline tools/run_pyright/compare.py]: nit: maybe we can utilize the `textwrap.indent` function for that?
Something like:
```python
lines = text.splitlines(keepends=True)
return lines[0] + textwrap.indent(''.join(lines[1:]), prefix, lambda_: True)
```

**Review by joeldierkes — DISMISSED** [review]
Good PR!

nit: The second commit message is not wrapped.
nit: The phrase `Pyright would have a hard time without any annotations` in the commit message is rather generic. Maybe it should be explained here that Pyright is more of a "static" type checker and requires some type annotations. While other type checkers are more dynamic and better at type inference, they did not provide the needed functionalities and integration into tooling. (or are just painfully slow).

File: tools/run_pyright/compare.py
Comment (joeldierkes) [inline tools/run_pyright/compare.py:73]: :joy:

**Comment by akselaase** [comment]
Slightly edited the commit message and simplified the indentation function. This dismissed the review, is this desired behaviour? It can be configured as shown here: https://github.com/isaacs/github/issues/1157#issuecomment-450409941.

**Comment by joeldierkes** [comment]
I don't know if I'n in favor of that. This use case hardly ever appears, and to resolve it you just have to re-request a new review.

Since we work on an open-source project, I'd rather have explicit approval than someone who changes more than the requested changes.

**Review by joeldierkes — DISMISSED** [review]

**Comment by akselaase** [comment]
To make sure this works in the development container, https://github.com/rucio/containers/pull/197 should also be merged. I can prepare a separate PR for documentation.

File: .github/workflows/autotest.yml
Comment (akselaase) [reply .github/workflows/autotest.yml]: Added `--Werror`. Better to get the reviewer to decide in case something is questionable.

**Comment by bari12** [comment]
Needs to be rebased due to an issue which existed on master yesterday.

**Comment by joeldierkes** [comment]
I think this is good to go :+1:

**Review by bari12 — APPROVED** [review]

**Pull request merged**
